src/Compilers/CSharp/Portable/Symbols/Synthesized/SynthesizedDelegateSymbol.cs

```diff
@@ -81,6 +81,11 @@ internal override NamedTypeSymbol BaseTypeNoUseSiteDiagnostics
             get { return ContainingAssembly.GetSpecialType(SpecialType.System_MulticastDelegate); }
         }
 
+        public override bool AreLocalsZeroed
+        {
+            get { return true; }
```

Comment:
> return true; [](start = 17, length = 13)

Is this method ever getting called?

Reply:
Yes. We have some tests that reach this method, most of them in `CodeGen_DynamicTests`. `GeneratedDelegates` and `GenericContainer2` are two examples.

Reply:
> Yes. We have some tests that reach this method, most of them in CodeGen_DynamicTests. GeneratedDelegates and GenericContainer2 are two examples.

It would be good to understand why this code is getting hit. Methods in delegates do not have bodies, why are we asking this question for them?

---
In reply to: [173008034](https://github.com/dotnet/roslyn/pull/25276#discussion_r173008034) [](ancestors = 173008034)

Reply:
> I took a look and two methods are being created in SynthesizedDelegateSymbol constructor: a DelegateConstructor and an InvokeMethod. Both of them have bodies. 

This doesn't sound right. What unit-test hits this code path? All delegate methods should be marked as  Runtime (0x0003 Method impl is provided by the runtime) and should have no bodies in metadata, including the constructor and Invoke. Either we are calling this API when there is no need to call it, or we are emitting incorrect metadata for the delegate type.

---
In reply to: [173031162](https://github.com/dotnet/roslyn/pull/25276#discussion_r173031162) [](ancestors = 173031162)

Reply:
A lot of unit tests hit it. `GeneratedDelegates` in `CodeGen_DynamicTests` is one of them.

Reply:
It looks like members of the synthesized delegate are properly marked and don't have bodies in metadata. I think we should adjust compiler's behavior to avoid calling this API for types and methods like this:
```
TypeDef #1 (02000002)
-------------------------------------------------------
 TypDefName: <>F{00000004}`4  (02000002)
 Flags     : [NotPublic] [AutoLayout] [Class] [Sealed] [AnsiClass]  (00000100)
 Extends   : 01000006 [TypeRef] System.MulticastDelegate
 4 Generic Parameters
  (0) GenericParamToken : (2a000001) Name : T1 flags: 00000000 Owner: 02000002
  (1) GenericParamToken : (2a000002) Name : T2 flags: 00000000 Owner: 02000002
  (2) GenericParamToken : (2a000003) Name : T3 flags: 00000000 Owner: 02000002
  (3) GenericParamToken : (2a000004) Name : TResult flags: 00000000 Owner: 02000002
 Method #1 (06000001) 
 -------------------------------------------------------
  MethodName: .ctor (06000001)
  Flags     : [Public] [HideBySig] [ReuseSlot] [SpecialName] [RTSpecialName] [.ctor]  (00001886)
  RVA       : 0x00000000
  ImplFlags : [Runtime] [Managed]  (00000003)
  CallCnvntn: [DEFAULT]
  hasThis 
  ReturnType: Void
  2 Arguments
   Argument #1:  Object
   Argument #2:  I
  2 Parameters
   (1) ParamToken : (08000001) Name : object flags: [none] (00000000)
   (2) ParamToken : (08000002) Name : method flags: [none] (00000000)

 Method #2 (06000002) 
 -------------------------------------------------------
  MethodName: Invoke (06000002)
  Flags     : [Public] [Virtual] [HideBySig] [NewSlot]  (000001c6)
  RVA       : 0x00000000
  ImplFlags : [Runtime] [Managed]  (00000003)
  CallCnvntn: [DEFAULT]
  hasThis 
  ReturnType: Var!3
  3 Arguments
   Argument #1:  Var!0
   Argument #2:  Var!1
   Argument #3:  ByRef Var!2
```

---
In reply to: [173270052](https://github.com/dotnet/roslyn/pull/25276#discussion_r173270052) [](ancestors = 173270052)